[jit] speed up alias analysis

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37106 [wip] update graph fuser aliasdb in-place
* **#36345 [jit] speed up alias analysis**

During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

suo committed:
[jit] speed up alias analysis

During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) at 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

[ghstack-poisoned]

eellison commented:
@suo could you look into the failures before I review? i would expect this shouldn't have any behavioral changes bc it's not changing the analysis just speeding it up (it shouldn't break tests).

IlyaOvodov commented:
@suo, 75% performance boosting is a great success, but still an application that starts for 3 minutes is almost as useless as one that starts for 10 min.
As far as I understood from reading issues and forums, this huge time is spent on optimization (that gives about several % of performance boost and is repeated every time when input shape is changed), compared with pure python PyTorch which starts the same net in less than 1 sec.
If I'm right, possibly it's a solution to give a way to switch this optimisation off when time it takes become unacceptable? Sorry if the reality is more complex than I expect :)

suo commented:
@IlyaOvodov: correct, the optimization time on large networks is still way too high. In production environments, we typically expect this time to be amortized over the lifetime of model serving, but even then the current state is bad.

There are a number of ways to tune the execution strategy. For example, if you don't want to optimization and basically want the code to run in interpreter-mode, you can set the following:
```
with torch.jit.optimized_execution(false):
    my_model(inputs)
```

IlyaOvodov commented:
@suo Wow! What a great undocumented cheat code! 
But is there a way to do the same (turn it off) in C++ inference?

suo commented:
> But is there a way to do the same (turn it off) in C++ inference?
answered on https://github.com/pytorch/pytorch/issues/36040 as it's more appropriate for discussion.

suo committed:
Update on "[jit] speed up alias analysis"


During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

[ghstack-poisoned]

eellison commented:
@suo i'm reviewing now but could you expand parts 2 and 3 in your summary with more information?
> Properly memoize/dynamic program the memory locations lookups.
how ?

> Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware setWildcards.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.
what is the naive solution and why doesn't it work, what is the new approach why does that solve it

suo commented:
> @suo i'm reviewing now but could you expand parts 2 and 3 in your summary with more information?
> 
> > Properly memoize/dynamic program the memory locations lookups.

The previous bfs implementation wasn't re-using the cached memorylocation lookup result, so we would re-do the whole search instead of memoizing.


> > Done naively, setting wildcards invalidates the above memoization,
> > trigger costly recomputation. So I added a cache-aware setWildcards.
> > Sadly that means you need alias analysis to reach into the guts of
> > memorydag, but the speedup is worth it.
> 
> what is the naive solution and why doesn't it work, what is the new approach why does that solve it

Naive solution is to invalidate everything in the transitive "points-from" closure to the wildcard. Since `setWildcard` only affects memorylocations (which are at the "source" of the points-from graph), it has the effect of invalidating basically the entire cache.

The new approach is commented in the code, but basically involves updating the cache in place in linear time.

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
+      auto contained_elem = memoryDAG_->fromIndex(loc);
+      // we only register writes on memory locations
+      if (contained_elem->pointsTo.empty()) {
+        writeIndex[write.first].set(contained_elem->index);
```
Why are we calling `.set()` here and `|=` above ? Could we be just be using `=` instead of `|=` above

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
+    }
+  }
+
+  for (const auto& write : writeRegistry_->writesToAllWildcards_) {
```
I'm not sure why we have `writesToAllWildcards`. It's only called on `aten::wait`, where we could instead just `analyzeConservative`.

suo replied:
the comment explains why. We cannot just be conservative and write to all inputs, because we don't know what the forked inputs are.

eellison replied:
If we analyzed `aten::fork` and `aten::wait` with `analyzeConservative` that would be correct, and would not require the above special casing leaking into memory dag

suo replied:
```
f : Future[None] = fork(a, b, c)  // mutates a
wait(f)
use(a)
```
it is illegal to reorder `wait` after `use`. `analyzeConservative` will only register a write to `f`, not `a`.

This can probably be improved. But I don't want to change behavior in this PR if possible.

eellison replied:
`analyzeConservative` marks a write to all contained elements:
```
    auto elem = elementMap_.at(input);
    registerWrite(input, node);
    MemoryLocations mem_locations;
    memoryDAG_->collectAllContainedMemoryLocations(elem, mem_locations);
    for (unsigned loc : mem_locations) {
      auto contained_elem = memoryDAG_->fromIndex(loc);
      // we only register writes on memory locations
      if (contained_elem->pointsTo.empty()) {
        registerWrite(contained_elem, node);
      }
    }
```

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -173,16 +241,13 @@ bool AliasDb::hasWriters(const Value* v) const {
     return false;
   }
```
Unrelated to this PR but: in `hasWriters` we're not testing the invariant that mutable values have an `Element *`

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -1409,38 +1467,21 @@ c10::optional<Element*> AliasDb::setWildcard(const Value* v) {
-  writeCache_ = {};
-  for (const auto& pr : writeIndex_) {
+MemoryLocations AliasDb::buildWriteCache() const {
+  MemoryLocations ret;
```
Why was this changed from mutating & void to functional and returning a new value?

suo replied:
Just because it is easier for me to reason about functional things

eellison commented on torch/csrc/jit/passes/utils/memory_dag.cpp:
```diff
@@ -93,71 +107,103 @@ bool MemoryDAG::mayContainAlias(
+    }
+  }
+
+  e->cachedMemoryLocations_ = std::move(ret);
```
What prevents this cache from getting invalidated ?

suo replied:
MemoryDAG is immutable (except through the specially marked `setWildcard` which is responsible for keeping the caches consistent).

eellison reviewed (commented):
I had a hard time reviewing this code for correctness. There are a lot of new caches, and it's hard to parse out what their lifetimes are and when they are invalidated etc. 

Could you please write up a description of how everything works together?

suo committed:
Update on "[jit] speed up alias analysis"


During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

[ghstack-poisoned]

suo commented on torch/csrc/jit/ir/alias_analysis.h:
```diff
@@ -233,11 +235,18 @@ class AliasDb {
-  mutable MemoryLocations writeCache_;
-  mutable bool isWriteCacheStale_ = true;
-  void rebuildWriteCache() const;
+  using TWriteIndex = ska::flat_hash_map<Node*, MemoryLocations>;
```
Type

suo commented on torch/csrc/jit/passes/utils/memory_dag.h:
```diff
@@ -69,22 +68,59 @@ class TORCH_API MemoryDAG {
+  //
+  // NOTE: This function consumes this builder, and any further mutation/usage
+  // is undefined.
+  std::unique_ptr<MemoryDAG> build();
```
good idea, will change

suo commented on torch/csrc/jit/passes/utils/memory_dag.h:
```diff
@@ -69,22 +68,59 @@ class TORCH_API MemoryDAG {
+ * Helper to build up the points-to graph.
+ *
+ * We separate the "building" into a different class because it allows us to
+ * cache internally to MemoryDAG without worrying about how the DAG structure
```
It means: during building we do not build caches. Then, when we're done, we can write caches is on an immutable object which is a lot simplier

suo commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
+    }
+    std::unordered_map<Node*, const Value*> writes_;
+    std::unordered_map<Node*, const Value*> containedWrites_;
+    std::unordered_set<Node*> writesToAllWildcards_;
```
explained on aten::wait discussion below

suo commented on torch/csrc/jit/ir/alias_analysis.h:
```diff
@@ -233,11 +235,18 @@ class AliasDb {
+  MemoryLocations buildWriteCache() const;
+
+  std::unordered_set<const Value*> wildcards_;
+  std::unordered_map<Element*, std::vector<const Value*>> wc_;
```
yeah actually this is totally redundant and unused but snuck in there :P I'll fix

suo commented on torch/csrc/jit/ir/alias_analysis.h:
```diff
@@ -233,11 +235,18 @@ class AliasDb {
+  c10::optional<TWriteIndex> writeIndex_;
+
+  struct WriteRegistry;
+  std::unique_ptr<WriteRegistry> writeRegistry_;
```
I will add in a comment:
- writeRegistry_ keeps track of writes as we are building aliasdb
- writeIndex_ is built from the `writeRegistry_` and can answer queries efficiently
- writeCache_ is a different index built from `writeRegistry_` that can answer a different kind of query. I'll rename them for clarity.
- wc_: is unused and should be deleted.

suo commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
+        it != elementMap_.end(), "Tried to write to value not in MemoryDAG");
+    const auto& writtenMemoryLocations =
+        memoryDAG_->getMemoryLocations(it->second);
+    writeIndex[write.first] |= writtenMemoryLocations;
```
write index is a cache from node -> all memory locations it writes to. This for loop builds it up based on information we recorded during analysis. I'll rewrite the code for clarity

suo commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -126,10 +126,78 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
+    void registerWriteToAllWildcards(Node* n) {
+      writesToAllWildcards_.insert(n);
+    }
+    std::unordered_map<Node*, const Value*> writes_;
```
yes good catch, this is a bug. I'll fix. 

containedWrites are separate because it produces more restrictive results (e.g. appending to a list does not write to elements in the list, but it writes to list itself)

suo committed:
Update on "[jit] speed up alias analysis"


During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

[ghstack-poisoned]

suo committed:
Update on "[jit] speed up alias analysis"


During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

[ghstack-poisoned]

suo committed:
Update on "[jit] speed up alias analysis"


During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

[ghstack-poisoned]

suo committed:
Update on "[jit] speed up alias analysis"


During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

[ghstack-poisoned]

suo committed:
Update on "[jit] speed up alias analysis"


During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -126,10 +126,81 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
 
+// Structure that keeps tracks of all writes at a high level. When alias
+// analysis is done, this will be used to construct a more efficient WriteIndex.
+struct AliasDb::WriteRegistry {
```
Maybe add comment somewhere that this exists before there is a value -> element mapping

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -126,10 +126,81 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
+
+  for (const auto& write : writeRegistry_->containedWrites_) {
+    Node* node = write.first;
+    const std::vector<const Value*> writtenValues = write.second;
```
nit: probably want `const std::vector<const Value*>&` right ?

suo replied:
good catch!

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -126,10 +126,81 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
+
+  for (const auto& write : writeRegistry_->writesToAllWildcards_) {
+    for (const auto& pr : wildcardIndex_) {
+      writeIndex[write].set(pr.second->index);
```
are you sure this api call is right `writeIndex[write].set(pr.second->index)` ? Why are we using `.set()` here instead of `|=` as above ?

suo replied:
|= is for adding many memorylocations. set() is for adding one.

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -126,10 +126,81 @@ bool AliasDb::isContainerType(const TypePtr& type) const {
+    }
+  }
+
+  writeRegistry_ = nullptr; // to make future access an error
```
maybe add some more comments here

eellison commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -779,17 +839,7 @@ void AliasDb::analyzeConservative(Node* node) {
-        registerWrite(contained_elem, node);
-      }
-    }
+    registerWrite(input, node, true);
```
nit can you add `/*writes_to_contained*/`  to the `true` arg.

eellison commented on torch/csrc/jit/passes/utils/memory_dag.cpp:
```diff
@@ -93,71 +107,101 @@ bool MemoryDAG::mayContainAlias(
-  if (!cachedMemoryLocations_.empty()) {
-    return cachedMemoryLocations_;
+const MemoryLocations& MemoryDAG::getMemoryLocations(const Element* e) const {
+  if (e->cachedMemoryLocations_) {
```
nit: can you add comment to the caching here: when you should expect it to be set

eellison commented on torch/csrc/jit/passes/utils/memory_dag.cpp:
```diff
@@ -93,71 +107,101 @@ bool MemoryDAG::mayContainAlias(
-    auto el = dag.fromIndex(index);
-    if (el->pointsTo.empty()) {
-      res.set(index);
+  if (e->pointsTo.empty()) {
```
It's a little surprising that we're not using a `seen` set in our DFS implementation, it may not trigger now but that's an easy way for an exponential runtime. Could you add a `getMemoryLocationsImpl` with a seen set ?

suo replied:
Not sure what a seen set would do? If we have traversed `e` before, its `cachedMemoryLocations_` will be populated and thus we will return immediately from the traversal.

eellison replied:
yea nevermind, cachedMemoryLocations_ makes sure we don't redo work.

eellison reviewed (approved):
This is a lot of complexity but given how much this is on the hot path i'm approving. Most of the complexity is because of wildcards and how they invalidate the memory dag state. We should investigate a context sensitive heap analysis so that wildcards are no longer special cased within the dag. 

Could you please address my comment in `getMemoryLocations` before landing ?

suo committed:
Update on "[jit] speed up alias analysis"


During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

[ghstack-poisoned]

suo committed:
Update on "[jit] speed up alias analysis"


During compilation, we spend a huge amount of time in alias analyis.
This PR does a few things to speed it up.

1. Separate the analysis into two phases: one where we build up the
necessary data structures, and the other where we service aliasing
queries. This allows us to defer building indices/maintaining index
consistency until after the "buildup" phase is done.

2. Properly memoize/dynamic program the memory locations lookups.

3. Done naively, setting wildcards invalidates the above memoization,
trigger costly recomputation. So I added a cache-aware `setWildcards`.
Sadly that means you need alias analysis to reach into the guts of
memorydag, but the speedup is worth it.

Sadly, these changes are kind of coupled for correctness reasons, so
they're all here at once.

I used this model (thanks @IlyaOvodov) as a provisional benchmark. You
can get it here:
https://www.dropbox.com/s/jlyygn6yygj1jkx/yolov3.zip. Unzip at run
`python test_timing.py`.

Baseline: (752.076s) right before 6bc8ffe82462c77ac4f9b27452046cb1f8f07d92
After optimizing before inlining: (699.593s)
After deferring cache construction: (426.180s)
After cache-aware `setWildcards`: (193.678s)

So a nice 75% speedup to overall compilation. There's a lot more to do
in other places of the compilation pipeline though.

Followup to this PR specifically:  Everything that fans out from the
`analyze` call is the "buildup" phase of AliasDB construction. This
should be factored into a separate analysis pass to statically
distinguish the two phases (right now we just null out stuff to
accomplish the same thing dynamically).

Differential Revision: [D20952727](https://our.internmc.facebook.com/intern/diff/D20952727)

[ghstack-poisoned]
